src/strategies/Staking.sol

```diff
@@ -4,174 +4,348 @@ pragma solidity =0.8.16;
-
-        // Unwrap all WETH
-        IWETH(address(WETH)).withdraw(WETH.balanceOf(address(this)));
+    enum Loan {
+        open,
+        close
     }
-
+  
     function receiveFlashLoan(
         ERC20[] memory tokens,
         uint256[] memory amounts,
         uint256[] memory, /* feeAmounts */
         bytes memory userData
     ) external override {
-        require(msg.sender == address(vault), "Staking: only vault");
+        require(msg.sender == address(balancer), "Staking: only balancer");
+
+        uint256 ethBorrowed;
+        uint daiBorrowed;
+        (Loan l) = abi.decode(userData, (Loan));
+
+        if (l == Loan.close) {
+            // We only flashloan dai in certain cases during divestment
+            if (tokens.length == 2) {
+                daiBorrowed = amounts[0];
+                ethBorrowed = amounts[1];
+            } else {
+                ethBorrowed = amounts[0];
+            }
+            _endPosition(ethBorrowed, daiBorrowed);
+        } else {
+            ethBorrowed = amounts[0];
+            _openPosition(ethBorrowed);
+        }
+        
+        // Payback loan
+        WETH.safeTransfer(address(balancer), ethBorrowed);
+        // We take slightly more from the user than they want to deposit (for slippage). If there's any left over send back to the user.
+        if (l == Loan.open) WETH.safeTransfer(address(vault), WETH.balanceOf(address(this)));
+
+        // Repay Dai loan
+        if (daiBorrowed > 0) DAI.safeTransfer(address(balancer), daiBorrowed);
+    }
+ 
+    function _openPosition(uint ethBorrowed) internal {
+        // Convert WETH to stEth
+        IWETH(address(WETH)).withdraw(ethBorrowed);
+        Address.sendValue(payable(address(LIDO)), ethBorrowed);
+
+        uint compoundDebtEth = _openCdp(ERC20(address(LIDO)).balanceOf(address(this)));
+
+        // Convert eth to weth in order to pay back balancer flash loan of weth
+        IWETH(address(WETH)).deposit{value: compoundDebtEth}();
+    }
+
+    uint public cdpId;
+
+    function _openCdp(uint amountWStEth) internal returns (uint compoundDebtEth){
+        cdpId = MAKER.open(ilk, address(this));
+        address urn = MAKER.urns(cdpId);
+
+        // Lock wsteth in maker
+        WSTETH_JOIN.join(urn, amountWStEth);
+
+        // Borrow at 58% collateral ratio
+        uint daiPrice = _getDaiPrice();
+        uint debt = _ethToDai(LIDO.getStETHByWstETH(amountWStEth), daiPrice).mulDivDown(58, 100);
+        MAKER.frob(cdpId, int(amountWStEth), int(debt)); 
+        
+        // Transfer Dai from cdp to this contract
+        MAKER.move({cdp: cdpId, dst: address(this), rad: debt * 1e27});
+        VAT.hope(address(joinDai));
+        joinDai.exit(address(this), debt);
+
+        // Deposit Dai in compound v2
+        cDAI.mint({underlying: debt});
 
-        ERC20 asset = tokens[0];
-        uint256 amount = amounts[0];
+        // Borrow at 75%
+        compoundDebtEth= _daiToEth(debt, daiPrice).mulDivDown(75, 100); 
```

Comment:
should we borrow 75% fixed or only the required amount to pay the loan?

Reply:
The 75% number is fixed. See the strategy doc.